crates/lib/src/bootc_kargs.rs

```diff
@@ -161,50 +168,56 @@ pub(crate) fn get_kargs(
     // A special case: if there's no kargs.d directory in the pending (fetched) image,
     // then we can just use the combined current kargs + kargs from booted
     if !fetched_tree.query_exists(cancellable) {
-        kargs.extend(existing_kargs);
+        kargs.extend(&existing_kargs);
         return Ok(kargs);
     }
 
     // Fetch the kernel arguments from the new root
     let remote_kargs = get_kargs_from_ostree(repo, &fetched_tree, sys_arch)?;
 
-    // get the diff between the existing and remote kargs
-    let mut added_kargs = remote_kargs
-        .clone()
-        .into_iter()
-        .filter(|item| !existing_kargs.contains(item))
-        .collect::<Vec<_>>();
-    let removed_kargs = existing_kargs
-        .clone()
-        .into_iter()
-        .filter(|item| !remote_kargs.contains(item))
-        .collect::<Vec<_>>();
+    // Calculate the diff between the existing and remote kargs
+    let added_kargs: Vec<_> = remote_kargs
+        .iter()
+        .filter(|item| !existing_kargs.iter().any(|existing| *item == existing))
+        .collect();
+    let removed_kargs: Vec<_> = existing_kargs
+        .iter()
+        .filter(|item| !remote_kargs.iter().any(|remote| *item == remote))
+        .collect();
 
     tracing::debug!(
         "kargs: added={:?} removed={:?}",
         &added_kargs,
         removed_kargs
     );
 
-    // apply the diff to the system kargs
-    kargs.retain(|x| !removed_kargs.contains(x));
-    kargs.append(&mut added_kargs);
+    // Apply the diff to the system kargs
+    for arg in &removed_kargs {
+        kargs.remove_exact(arg);
+    }
+    for arg in &added_kargs {
+        kargs.add(arg);
+    }
 
     Ok(kargs)
 }
 
 /// This parses a bootc kargs.d toml file, returning the resulting
 /// vector of kernel arguments. Architecture matching is performed using
 /// `sys_arch`.
-fn parse_kargs_toml(contents: &str, sys_arch: &str) -> Result<Vec<String>> {
+fn parse_kargs_toml(contents: &str, sys_arch: &str) -> Result<Option<Cmdline<'static>>> {
```

Comment:
It confused me at first that this is using `'static` but I guess it means we're using owned values. Maybe worth a `CmdlineOwned` alias?